compiler/backend/cgir.nim

```diff
@@ -0,0 +1,203 @@
+    # ``x = hiddenAddr y`` operation
+    # For views, consider lowering them into pointers (for backends where it
+    # makes sense, e.g. C). This could make ``cnkDerefView`` obsolete
+
+    cnkConv          ## a type conversion
+    # future direction: introduce a dedicated operation for "l-value preserving
+    # conversions"
+    cnkHiddenConv
+    # future direction: the notion of "hidden" doesn't make any sense in the
+    # context of code generation. Adjust the code generators so that they no
+    # longer depend on ``cnkHiddenConv`` being different from ``cnkConv``, and
+    # then remove the former
+
+    cnkObjDownConv   ## down conversion between `object` or `ref` types
+    cnkObjUpConv     ## up conversion between `object` or `ref` types
+
+    cnkCast          ## reinterpret the bit-pattern of the operand as a
+                     ## different type
+
+    # ---- special conversions kept for compatibility
+    cnkChckRangeF      ## range check for floats
+    cnkChckRange64     ## range check for 64-bit ints
+    cnkChckRange       ## range check for ints
+    cnkStringToCString ## string to cstring
+    cnkCStringToString ## cstring to string
+    # future direction: lower these coversion operations during the MIR
+    # phase and then remove the node kinds
+    # ---- end
+
+    cnkStmtList
+    cnkStmtListExpr
+    # future direction: remove ``cnkStmtListExpr``. The code generators know
+    # based on the context a statement list appears in whether its an
+    # expression or not
+
+    cnkVoidStmt   ## discard the operand value (i.e., do nothing with it)
+    cnkPragmaStmt ## a single compiler directive
+    cnkEmitStmt   ## an ``emit`` statement
+    cnkAsmStmt    ## an ``asm`` statement
+
+    cnkIfStmt     ## only execute the body when the condition expression
+                  ## evaluates to 'true'
+    cnkRepeatStmt ## execute the body indefinitely
+    cnkCaseStmt   ## a ``case`` statement
+    cnkBlockStmt  ## an (optionally) labeled block
+
+    cnkBreakStmt  ## break out of labeled block, or, if no label is provided,
+                  ## the closest ``repeat`` loop
+    cnkRaiseStmt  ## raise(x) -- set the `x` as the current exception and start
+                  ## exceptional control-flow. `x` can be ``cnkEmpty`` in which
+                  ## case "set current exception" part is skipped
+    # future direction: lower the high-level raise statements (which means
+    # "set the current exception" + "start exceptional control-flow") into
+    # just "start exceptional control-flow"
+    cnkReturnStmt
+
+    cnkTryStmt
+    cnkExcept
+    cnkFinally
+
+    cnkBranch     ## the branch of a ``case`` statement
+
+    cnkDef        ## starts the lifetime of a local and optionally assigns an
+                  ## initial value
+
+    cnkAsgn       ## a = b
+    cnkFastAsgn   ## fast assign b to a
+    # future direction: have ``cnkAsgn`` mean "assign without implying any
```

Comment:
That's a good question. Everything related to assignment is very fuzzy at the moment, mainly because of "assignment" having different meaning depending on the used code generator (they don't operate on the same language level with regards to assignments).

Broadly speaking, `cnkAsgn` combines both "assign" (copy to non-empty destination) *and* "initialize" (copy to empty destination), while `cnkFastAsgn` means "create a shallow copy". Whether combining "assign" and "initialize" is a good idea, I'm not sure (it probably isn't), but `PNode` did it, so for initial compatibility, I carried it over.

Reply:
Sounds like we're on the same page, needs some more thinking.